File: Leetcode Daily Challenge/025-July-2025/006-finding-pairs-with-a-certain-sum.java
// https://leetcode.com/problems/finding-pairs-with-a-certain-sum/




import java.util.*;

class FindSumPairs {
    private int[] a, b;
    private Map<Integer, Integer> mpp;

    public FindSumPairs(int[] nums1, int[] nums2) {
        this.a = nums1;
        this.b = nums2;
        this.mpp = new HashMap<>();

        for (int num : nums2) {
            mpp.put(num, mpp.getOrDefault(num, 0) + 1);
        }    
    }
    
    public void add(int index, int val) {
        int original = b[index];
        b[index]    += val;
        mpp.put(original, mpp.get(original) - 1);
        mpp.put(b[index], mpp.getOrDefault(b[index], 0) + 1);
    }
    
    public int count(int tot) {
        int ans = 0;
        for (int num : a) {
            int tar = (tot - num);
            ans += mpp.getOrDefault(tar, 0);
        }
        return ans;
    }
}

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * FindSumPairs obj = new FindSumPairs(nums1, nums2);
 * obj.add(index,val);
 * int param_2 = obj.count(tot);
 */

File: Leetcode Daily Challenge/025-July-2025/006-finding-pairs-with-a-certain-sum.cpp
// https://leetcode.com/problems/finding-pairs-with-a-certain-sum/




#include <bits/stdc++.h>
using namespace std;

class FindSumPairs {
public:
    vector<int> a, b;
    unordered_map<int, int> mpp;

    // TC : O(N) for initialization
    // SC : O(N) for storing frequency of nums2 elements
    FindSumPairs(vector<int>& nums1, vector<int>& nums2) {
        a = nums1;
        b = nums2;
        for (int &num : nums2) {
            mpp[num] ++;
        }
    }
    
    // TC : O(1)
    void add(int index, int val) {
        int original = b[index];
        b[index] += val;
        mpp[original] --;
        mpp[original + val] ++;
    }
    
    // TC : O(N) for each count call
    int count(int tot) {
        int ans = 0;
        for (int &val : a) {
            int tar = (tot - val);
            if (mpp.count(tar)) {
                ans += mpp[tar];
            }
        }
        return ans;
    }
};

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * FindSumPairs* obj = new FindSumPairs(nums1, nums2);
 * obj->add(index,val);
 * int param_2 = obj->count(tot);
 */